### PR TITLE
gradle: add toolchain paths to gradle.properties

### LINKED ISSUE #207153: `gradle`'s toolchain support does not work with IntelliJ IDEA
### Describe the bug

When using the `gradle` with the `javaToolchains` option, toolchains are set up in a way that is not compatible with IntelliJ IDEA or any other tool that does not make use of a binary to launch Gradle. When using the built Gradle instance in IntelliJ IDEA, IDEA does not use the chosen toolchain settings at all.

### Steps To Reproduce

1. Create a directory with the following files:

#### shell.nix

```nix
{ pkgs ? import <nixpkgs> { } }:

let
  unstable = import (fetchTarball https://github.com/NixOS/nixpkgs/archive/nixos-unstable.tar.gz) { };
in

pkgs.mkShell {
  nativeBuildInputs = [
    (unstable.gradle.override {
      javaToolchains = [ unstable.jdk11 unstable.jdk17 ];
    })
  ];
}
```

#### settings.gradle

```gradle
include 'proj1', 'proj2'
```

#### proj1/build.gradle

```gradle
plugins {
	id 'java'
}

java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(11)
    }
}
```

#### proj1/src/main/java/Hello.java

```java
class Hello {
    public static void main(String[] args) {
        System.out.println("Hello, world!");
    }
}
```
#### proj2/build.gradle

```gradle
plugins {
	id 'java'
}

java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(17)
    }
}
```

#### proj2/src/main/java/Hello.java

```java
class Hello {
    public static void main(String[] args) {
        System.out.println("Hello, world!");
    }
}
```

(example repository TODO)

2. Enter a nix shell (`nix-shell`)
3. Run `gradle build`. Everything works.
4. Run `which gradle` to retrieve the path to the gradle install in the nix store, e.g. `/nix/store/abcd-gradle/bin/gradle`
5. Open IntelliJ IDEA (e.g. `jetbrains.idea-community`) to the created directory. At this point, importing will fail in any case because IDEA will launch a "vanilla" Gradle that provisions non-Nix-compatible JDKs.
6. Change the Gradle install directory to the one found with `which`: open settings (`Ctrl+Alt+S`), go to "Build, Execution, Deployment" -> "Build Tools" -> "Gradle". In the "Use Gradle from" field, choose "Specified location" and input `/nix/store/abcd-gradle/lib/gradle` (note the *lib* instead of *bin*)
7. At this point, the sync will fail again with the same error

The errors will usually be something about a JDK not being probed due to a "file not found" error.

### Expected behavior

IntelliJ IDEA is able to import the project because we pointed it to the correct `gradle` installation path.

### Context and workaround

For more context, this seems to be because IntelliJ IDEA does not use the executable binary (which sets up toolchains via environment variables) at all. However, both IntelliJ IDEA and the executable will pick up [Gradle properties](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties) specified in a `gradle.properties` file in the "Gradle install directory".  In our case, this special directory is `/nix/store/abcd-gradle/lib/gradle`, so we can put a custom file in `/nix/store/abcd-gradle/lib/gradle/gradle.properties`

TL;DR:

|  | CLI | IntelliJ IDEA |
|---|---|---|
| Wrapper executable | Works | Does not work |
| `gradle.properties` | Works | Works |

You can use the following `shell.nix` as a workaround:

```nix
{ pkgs ? import <nixpkgs> { } }:

let
  unstable = import (fetchTarball https://github.com/NixOS/nixpkgs/archive/nixos-unstable.tar.gz) { };
  toolchains = [ unstable.jdk11 unstable.jdk17 ];
in

pkgs.mkShell {
  nativeBuildInputs = [
    (unstable.gradle.overrideAttrs (curr: old: {
      fixupPhase = old.fixupPhase + ''
        cat > $out/lib/gradle/gradle.properties <<EOF
        org.gradle.java.installations.paths=${pkgs.lib.concatStringsSep "," toolchains}
        EOF
      '';
    }))
  ];
}
```

Note that here we're using a variable in a `let ... in` block, but the package could be adapted to use the existing `javaToolchains` option.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

I don' tthink this is related but here you go:

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.83, NixOS, 22.11 (Raccoon), 22.11.1008.0938d73bb14`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos-22.11"`
 - channels(matthieu): `"home-manager-22.11.tar.gz, nixos-unstable"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by addict3d**
This way IntelliJ IDEA uses the overridden javaToolchains.

Fix #207153

I ran the package internal tests (`nix-build --attr pkgs.gradle.passthru.tests`) and verified success with IDEA.

```
nix-repl> :b local.pkgs.gradle.override { javaToolchains = [ pkgs.openjdk21 pkgs.openjdk23 ]; }

This derivation produced the following outputs:
  out -> /nix/store/j7b4bki0x0yswanah8l78xgad0c12i8d-gradle-8.10.2
```
the resulting change looks like this:
```
$ cat /nix/store/j7b4bki0x0yswanah8l78xgad0c12i8d-gradle-8.10.2/lib/gradle/gradle.properties 

org.gradle.java.installations.paths=/nix/store/f042x32jfm94d3cgaga8d6xl8vy6sg46-openjdk-21.0.5+11,/nix/store/wkyv98s8q4wqdzpjd34z4hk0lk8fmww5-openjdk-23.0.1+11
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/tools/build-managers/gradle/default.nix [hunk 1]
```diff
@@ -100,6 +103,8 @@ rec {
           makeWrapper ${java}/bin/java $out/bin/gradle \
             ${varDefs}
             --add-flags "-classpath $gradle_launcher_jar org.gradle.launcher.GradleMain${toolchain.property}"
+
+          echo "${toolchain.paths}" > $out/lib/gradle/gradle.properties
```
Comment (liff) [inline pkgs/development/tools/build-managers/gradle/default.nix:107]: Should this be conditional on `javaToolchains` being non-empty? That is, if it’s empty, we have a `gradle.properties` containing `org.gradle.java.installations.paths=`. It doesn’t seem like that _should_ cause problems, but can it?

Comment (addict3d) [reply pkgs/development/tools/build-managers/gradle/default.nix:107]: Good question. Am happy to make this conditional if it seems better or less problematic. Presumably, with `javaToolchains` empty, the JDK would be specified elsewhere. I'll poke at gradle's behavior with `javaToolchains` empty.

With an empty list for `javaToolchains` currently, the resulting binary wrapper will be passing an empty string for `-Porg.gradle.java.installations.fromEnv=''`.

Comment (addict3d) [reply pkgs/development/tools/build-managers/gradle/default.nix:107]: I've verified that `gradle build` works with `javaToolchains` empty for a project that specifies a trivially available Java toolchain (set to 21, which is what Gradle currently has for itself).

Comment (liff) [reply pkgs/development/tools/build-managers/gradle/default.nix:107]: If everything works either way, I suppose we won’t need the condition.

**Review by liff — APPROVED** [review]

**Comment by addict3d** [comment]
A fresh rebase to fix the merge conflict.

File: pkgs/development/tools/build-managers/gradle/default.nix [hunk 2]
```diff
@@ -90,11 +90,14 @@ rec {
       installPhase =
         with builtins;
         let
+          # set toolchains via environment properties in wrapper and in gradle.properties.
+          # See https://docs.gradle.org/current/userguide/toolchains.html#sec:custom_loc
           toolchain = rec {
             prefix = x: "JAVA_TOOLCHAIN_NIX_${toString x}";
             varDefs = (lib.imap0 (i: x: "${prefix i} ${x}") javaToolchains);
             varNames = lib.imap0 (i: x: prefix i) javaToolchains;
             property = " -Porg.gradle.java.installations.fromEnv='${concatStringsSep "," varNames}'";
```
Comment (britter) [inline pkgs/development/tools/build-managers/gradle/default.nix:99]: Why are we keeping toolchains from env? Can't we drop this way of configuring toolchains now?

Comment (addict3d) [reply pkgs/development/tools/build-managers/gradle/default.nix:99]: The encoding as `JAVA_TOOLCHAIN_NIX_` does not seem to be documented as a user-facing aspect of the gradle wrapper.

We still want to set `JAVA_HOME` in the wrapper, right?

**Review by britter — APPROVED** [review]
Great stuff! I think we don't need to go throw the properties file and could instead pass the toolchains directly as a system property using `-Dorg.gradle.java.installations.paths=...`, but that could also be a follow up PR.

I like that you went the extra mile ensuring that this works by reconfiguring the tests. This made me aware that we don't have any coverage testing, that toolchains actually work. I will look into this tomorrow.

**Comment by addict3d** [comment]
> Great stuff! I think we don't need to go throw the properties file and could instead pass the toolchains directly as a system property using `-Dorg.gradle.java.installations.paths=...`, but that could also be a follow up PR.
> 
> I like that you went the extra mile ensuring that this works by reconfiguring the tests. This made me aware that we don't have any coverage testing, that toolchains actually work. I will look into this tomorrow.

Hrmm. Dropping the properties file will break usage from IntelliJ, as noted in https://github.com/NixOS/nixpkgs/issues/207153

**Comment by britter** [comment]
@addict3d okay, then let's merge this in as is.

**Comment by britter** [comment]
@addict3d would you mind rebasing this on lasted `master`?

**Comment by britter** [comment]
Merging this will break the test introduced in #391822 I need to rewrite the test to now rely on the output of the `javaToolchains` task. Then it should be save to merge this change.

**Comment by addict3d** [comment]
> Merging this will break the test introduced in #391822 I need to rewrite the test to now rely on the output of the `javaToolchains` task. Then it should be save to merge this change.

Okay, that looks like it fails due to the source changing. A patch like this should resolve it?
```
--- a/pkgs/development/tools/build-managers/gradle/default.nix
+++ b/pkgs/development/tools/build-managers/gradle/default.nix
@@ -302,9 +302,8 @@ rec {
                     env GRADLE_USER_HOME=$TMPDIR/gradle org.gradle.native.dir=$TMPDIR/native \
                       gradle javaToolchains --no-daemon --quiet --console plain > $out
                     cat $out | grep "Language Version:   11"
-                    cat $out | grep "Detected by:        environment variable 'JAVA_TOOLCHAIN_NIX_0'"
                     cat $out | grep "Language Version:   23"
-                    cat $out | grep "Detected by:        environment variable 'JAVA_TOOLCHAIN_NIX_1'"
+                    cat $out | grep "Detected by:        Gradle property 'org.gradle.java.installations.paths'"
                   '';
             } // gradle.tests;
           }
```

Or perhaps we want to be more specific and check something like `[[ 2 -eq $(cat out | grep -c "Detected by:        Gradle property 'org.gradle.java.installations.paths'") ]];`?

**Comment by britter** [comment]
Hey @addict3d I've changed the test to be more lenient to changes in how we set up toolchains: https://github.com/NixOS/nixpkgs/pull/395965 once that is merged the test should be passing even with your changes. I'll let you know once it is merged.

**Comment by britter** [comment]
@addict3d I finally got the changes to the new toolchains test merged. If you rebase on current master the test should be green even with your changes.

**Comment by addict3d** [comment]
> @addict3d I finally got the changes to the new toolchains test merged. If you rebase on current master the test should be green even with your changes.

@britter Rebased it. When I run the package tests, I see this error:

```
building '/nix/store/3bgvf653qw616jgrz64k1srwi9vippys-detects-toolchains-from-nix-env.drv'...

building '/nix/store/s61i71fvrv14mphypqj0i6gnjr3rk6wm-equal-contents-can-build-and-run-a-trivial-java-application.drv'...
Checking:
can build and run a trivial Java application
expected /nix/store/ndkyj3146ybnzpg3fri3kmhh31pfis01-expected and actual /nix/store/2zxn7agbc2wrj55n1hx7lv95dz30l5a6-actual match.
OK

error: builder for '/nix/store/3bgvf653qw616jgrz64k1srwi9vippys-detects-toolchains-from-nix-env.drv' failed with exit code 1;
       last 1 log lines:
       > 
       For full logs, run 'nix log /nix/store/3bgvf653qw616jgrz64k1srwi9vippys-detects-toolchains-from-nix-env.drv'.
```

It seems like the check compares the full version string from the java.version property with only the major version:

`test 23.0.2 = 23`

**Comment by britter** [comment]
@addict3d looks like the test was successful on hyra.

**Comment by addict3d** [comment]
> @addict3d looks like the test was successful on hyra.

Where is that? Still looks broken the same way.

**Comment by addict3d** [comment]
On a random machine:

```
nix-repl> lib.getVersion pkgs.openjdk23
"23.0.1+11"
```
Which in nix `lib.versions` speak you might call `major.minor.patch+build`


The `java.version` property is said to be the 'JRE version number', which since JDK 10 (ref 1) conform to this format: https://docs.oracle.com/en/java/javase/24/docs/api/java.base/java/lang/Runtime.Version.html#verStr

(ref 1: https://docs.oracle.com/en/java/javase/11/install/version-string-format.html)

So, shouldn't the 'detects-toolchains' compare the `major.minor.patch` from the nix jdk package against the output of `System.out.println(System.getProperty("java.version"));`?

**Comment by britter** [comment]
You're absolutely right in saying that this test is currently broken on master. Not sure how that happened 😳 Here's the fix: https://github.com/NixOS/nixpkgs/pull/403620

**Comment by britter** [comment]
@addict3d #403620 was merged. This one should finally be good after another rebase. Sorry for the hassle.

**Comment by addict3d** [comment]
@britter No worries, thank you!

**Review by britter — APPROVED** [review]

**Comment by britter** [comment]
@FliegendeWurst can we merge this?

**Review by FliegendeWurst — APPROVED** [review]
